### PR TITLE
Document JSON module

### PR DESCRIPTION
**Pull request by benharsh**
This PR adds documentation to the serializer API methods in the JSON module, and documents the way that serializer type-kinds are translated to valid JSON.

### CODE REVIEW

File: modules/standard/JSON.chpl [hunk 1]
```diff
@@ -63,6 +63,37 @@ module JSON {
...
+      would be serialized as a string: ``"red"``.
+
+      The ``nil`` value and nilable class variables storing ``nil`` will be
+      serialized as the text ``null``.
+
+      Classes and records will have their ``serialize`` method invoked, passing
+      in ``writer`` and this Serializer as arguments. Please see the
+      :ref:`serializers technote<ioSerializers>` for more.
+
+      Classes and records are expected to implement the ``writeSerializable``
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:91]: Should this be adjusted to read as on: https://github.com/chapel-lang/chapel/pull/23502/files#diff-e54b013aa9de867488474be4a98980482efcc0038d15722b3796c7276913a84bR3201 ?

File: modules/standard/JSON.chpl [hunk 2]
```diff
@@ -80,13 +111,65 @@ module JSON {
...
+      writer.writeLiteral("{");
+      return new AggregateSerializer(writer, _ending="}");
+    }
+
+    /*
+      Returned by ``startClass`` or ``startRecord`` to provide the API for
+      serializing classes or records.
+
+      Classes and records are serialized as JSON objects. For example, a
+      ``class`` or ``record`` with integer fields 'x' and 'y' with values '0'
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:147]: Should this be ``x``... ``0`` to be consistent with formatting for other examples?

File: modules/standard/JSON.chpl [hunk 3]
```diff
@@ -80,13 +111,65 @@ module JSON {
...
+      @chpldoc.nodoc
       const _ending : string;
+      @chpldoc.nodoc
       var _firstPtr : c_ptr(bool) = nil;
 
+      /*
+        Serialize ``field`` named ``name``.
+
+        Serializes fields in the form '"<name>":<field>'. Adds a comma before the
+        name if this is no the first field.
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:171]: no -> not

File: modules/standard/JSON.chpl [hunk 4]
```diff
@@ -80,13 +111,65 @@ module JSON {
...
       var _first : bool = true;
+      @chpldoc.nodoc
       const _ending : string;
+      @chpldoc.nodoc
       var _firstPtr : c_ptr(bool) = nil;
 
+      /*
+        Serialize ``field`` named ``name``.
+
+        Serializes fields in the form '"<name>":<field>'. Adds a comma before the
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:170]: Maybe double backtick the ``"<name>":<field>`` or use some other formatting techniques to make it pop.

File: modules/standard/JSON.chpl [hunk 5]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
+            var y : int;
+          }
+
+        would be serialized as:
+
+        .. code-block:: text
+
+          {"x":5, "y":2}
+
+        :arg writer: The ``fileWriter`` to be used when serializing. Must match
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:204]: -> Must match the?

File: modules/standard/JSON.chpl [hunk 6]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
+        :returns: A new AggregateSerializer
+      */
       proc ref startClass(writer, name: string, size: int) throws {
         return new AggregateSerializer(writer, _parent=true,
                                        _firstPtr=c_addrOf(_first));
       }
 
-      @chpldoc.nodoc
+      /*
+        Ends serialization of the current class by writing the character ``}``
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:217]: Ending `.`

File: modules/standard/JSON.chpl [hunk 7]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
       proc endClass() throws {
         if !_parent then
           writer._writeLiteral(_ending);
         else if _firstPtr != nil then
           _firstPtr.deref() = _first;
       }
 
-      @chpldoc.nodoc
+      /*
+        Ends serialization of the current record by writing the character ``}``
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:227]: Ending `.`

File: modules/standard/JSON.chpl [hunk 8]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
+        .. code-block:: text
+
+          {"x":5, "y":2}
+
+        :arg writer: The ``fileWriter`` to be used when serializing. Must match
+          writer used to create current AggregateSerializer.
+        :arg name: The name of the class type.
+        :arg size: The number of fields in the class.
+
+        :returns: A new AggregateSerializer
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:209]: A new :type:`AggregateSerializer`

File: modules/standard/JSON.chpl [hunk 9]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
+    /*
+      Start serializing a tuple by writing the character ``[``.
+
+      In this format tuples are serialized as JSON lists. For example, the
+      tuple ``(1, 2, 3)`` would be serialized as ``[1, 2, 3]``.
+
+      :arg writer: The ``fileWriter`` to be used when serializing.
+      :arg size: The number of elements in the tuple.
+
+      :returns: A new ListSerializer
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:243]: Link to the type?

File: modules/standard/JSON.chpl [hunk 10]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
-    proc startRecord(writer: jsonWriter, name: string, size: int) throws {
-      writer.writeLiteral("{");
-      return new AggregateSerializer(writer, _ending="}");
+    /*
+      Start serializing a list by writing the character ``[``.
+
+      :arg writer: The ``fileWriter`` to be used when serializing.
+      :arg size: The number of elements in the list.
+
+      :returns: A new ListSerializer
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:256]: Link to the type?

File: modules/standard/JSON.chpl [hunk 11]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
+      Start serializing an array.
 
-    @chpldoc.nodoc
-    proc startList(writer: jsonWriter, size: int) throws {
-      writer.writeLiteral("[");
-      return new ListSerializer(writer);
+      :arg writer: The ``fileWriter`` to be used when serializing.
+      :arg size: The number of elements in the array.
+
+      :returns: A new ArraySerializer
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:317]: Link to type?

File: modules/standard/JSON.chpl [hunk 12]
```diff
@@ -96,85 +179,185 @@ module JSON {
...
+
+      .. code-block:: text
+
+        [
+         [0, 1, 2],
+         [3, 4, 5],
+         [6, 7, 8]
+        ]
+
+      The indenting of these multidimensional arrays is not considered stable.
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:343]: Maybe put this in a warning block.

File: modules/standard/JSON.chpl [hunk 13]
```diff
@@ -249,29 +434,77 @@ module JSON {
...
-    @chpldoc.nodoc
-    proc startArray(writer: jsonWriter, size: int) throws {
-      return new ArraySerializer(writer);
+    /*
+      Start serializing a map by writing the character ``{``.
+
+      :arg writer: The ``fileWriter`` to be used when serializing.
+      :arg size: The number of entries in the map.
+
+      :returns: A new MapSerializer
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:462]: Link to type?

File: modules/standard/JSON.chpl [hunk 14]
```diff
@@ -249,29 +434,77 @@ module JSON {
...
+      .. code-block:: text
+
+        {
+          "east": "west",
+          "hello": "goodbye",
+          "north": "south",
+          "day": "night"
+        }
+
+      The indenting and whitespace of this example is not considered stable.
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:484]: Make a warning or have a warning at the top of the module about whitespace?

File: modules/standard/JSON.chpl [hunk 15]
```diff
@@ -249,29 +434,77 @@ module JSON {
...
+      @chpldoc.nodoc
       var writer;
+      @chpldoc.nodoc
       var _first : bool = true;
 
-      @chpldoc.nodoc
+      /*
+        Serialize ``key``.
+
+        JSON itself only supports strings as names in objects. This library
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:496]: This seems like an explanation that could be in a `::note`, coupled with the above.

File: modules/standard/JSON.chpl [hunk 16]
```diff
@@ -249,29 +434,77 @@ module JSON {
...
+    */
     record MapSerializer {
+      @chpldoc.nodoc
       var writer;
+      @chpldoc.nodoc
       var _first : bool = true;
 
-      @chpldoc.nodoc
+      /*
+        Serialize ``key``.
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:494]: Maybe "Serialize ``key`` and store it in an escaped JSON string."

File: modules/standard/JSON.chpl [hunk 17]
```diff
@@ -298,31 +531,22 @@ module JSON {
...
+        Serialize ``val``, preceded by the character ``:``.
+      */
       proc writeValue(const val: ?) throws {
         writer._writeLiteral(": ");
         writer.write(val);
       }
 
-      @chpldoc.nodoc
+      /*
+        Ends serialization of the current map by writing the character ``}``
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:543]: Ending `.`

File: modules/standard/JSON.chpl [hunk 18]
```diff
@@ -520,12 +837,23 @@ module JSON {
           reader.revert();
       }
 
+      /*
+        Start deserializing a nested class inside the current class.
+
+        See ``jsonSerializer.AggregateSerializer.startClass`` for details
+        on inheritance on the JSON format.
+
+        :returns: A new AggregateDeserializer
```
Comment (dlongnecke-cray) [inline modules/standard/JSON.chpl:846]: Link to type?

**Review by dlongnecke-cray — COMMENTED** [review]
I'd do a double check for `:arg:` descriptions and make sure they all consistently end with a period (or not). I can't remember what the style guide says about that, (my gut says they don't end with a period, but...).

Other than that, double check that links to the serializer types are all set up everywhere.

File: modules/standard/JSON.chpl (hunk 1 above)
Comment (benharsh) [reply modules/standard/JSON.chpl:91]: I think this shorter version works fine for the two options. With deserialization it felt worth a longer sentence to be clear.

File: modules/standard/JSON.chpl (hunk 8 above)
Comment (benharsh) [reply modules/standard/JSON.chpl:209]: There's actually a chpldoc bug with this one, unfortunately, so the ``:type`` won't work.

File: modules/standard/JSON.chpl (hunk 12 above)
Comment (benharsh) [reply modules/standard/JSON.chpl:343]: I was doing that initially, but unfortunately there's a chpldoc bug where the warning overflows and includes all the type's methods. The same goes for the code block preceding this, so having the text here is blocking that issue as well.

File: modules/standard/JSON.chpl (hunk 16 above)
Comment (benharsh) [reply modules/standard/JSON.chpl:494]: I think the full note is needed with examples, but maybe I can say:

```
Serialize ``key`` as the name of the member pair.
```

?

**Review by dlongnecke-cray — APPROVED** [review]

**Pull request merged**